relayer/strategies.go

```diff
@@ -32,62 +31,83 @@ func relayerMainLoop(ctx context.Context, log *zap.Logger, src, dst *Chain, filt
-				} else {
-					errCh <- fmt.Errorf("error querying all channels on chain{%s}@connection{%s}: %w",
-						src.ChainID(), src.ConnectionID(), err)
-				}
-				return
-			}
-
-			// Apply the channel filter rule (i.e. build allowlist, denylist or relay on all channels available)
-			srcChannels = applyChannelFilterRule(filter, srcChannels)
-
-			// Filter for open channels that are not already in our slice of open channels
-			srcOpenChannels = filterOpenChannels(srcChannels, srcOpenChannels)
+	// Query the list of channels on the src connection.
+	srcChannels, err := queryChannelsOnConnection(ctx, src)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			errCh <- err
+		} else {
+			errCh <- fmt.Errorf("error querying all channels on chain{%s}@connection{%s}: %w",
+				src.ChainID(), src.ConnectionID(), err)
 		}
+		return
+	}
 
+	// Apply the channel filter rule (i.e. build allowlist, denylist or relay on all channels available),
+	// then filter out only the channels in the OPEN state.
+	srcChannels = applyChannelFilterRule(filter, srcChannels)
+	srcOpenChannels := filterOpenChannels(srcChannels)
+
+	for {
 		// TODO once upstream changes are merged for emitting the channel version in ibc-go,
 		// we will want to add back logic for finishing the channel handshake for interchain accounts.
 		// Essentially the interchain accounts module will initiate the handshake and then the relayer finishes it.
 		// So we will occasionally query recent txs and check the events for `ChannelOpenInit`, at which point
 		// we will attempt to finish opening the channel.
 
+		// TODO when we have completed the Chain/Path processor refactor we will be listening for
+		// new channels that are being opened/created so it's possible we have no open channels to relay on
+		// at startup but after some time has passed a channel needs opened and relayed on. At this point we
+		// could choose to loop here until some action is needed.
 		if len(srcOpenChannels) == 0 {
-			continue
+			errCh <- fmt.Errorf("there are no open channels to relay on, check your config's allow/deny list rules")
+			return
 		}
 
-		var wg sync.WaitGroup
-		// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against
+		// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against.
 		for _, channel := range srcOpenChannels {
 			if !channel.active {
 				channel.active = true
-				wg.Add(1)
-				go relayUnrelayedPacketsAndAcks(ctx, log, &wg, src, dst, maxTxSize, maxMsgLength, channel, channels)
+				go relayUnrelayedPacketsAndAcks(ctx, log, src, dst, maxTxSize, maxMsgLength, channel, channels)
 			}
 		}
-		wg.Wait()
 
-		for channel := range channels {
-			channel.active = false
-			break
+		// Block here until one of the running goroutines exit.
+		channel, ok := <-channels
+		if !ok {
+			errCh <- fmt.Errorf("the (chan *ActiveChannel) was closed and no more values can be written/read")
```

Comment:
Now I see that this function creates `channels` and nobody else closes it. So it should be okay to drop the `, ok` check here, and drop deferring the channel close on L34.

The next question is, what happens if the main loop is blocking on this channel read and the main context is canceled?

That is, this could change to:

```go
var channel *ActiveChannel
select {
  case channel = <-channels:
    // Keep going.
  case <-ctx.Done():
    // Probably want to wait here for background goroutines to finish,
    // which means restoring the wait group.
    // The context cancelation should quickly propagate to all 
    // relayUnrelayedPacketsAndAcks goroutines.
    wg.Wait()
    errCh <- ctx.Err()
    return
}

channel.active = false
// ...
```

Reply:
Good catch there, went ahead and made the necessary changes so we are handling the main context being cancelled properly.